### ISSUE TITLE
Reorder ata_pm_standby and ata_apm_level commands for TOSHIBA HDD

### ISSUE THREAD

**Issue by crazygolem** [storaged-project/udisks#694]
I recently bought a Toshiba X300 HDD and had trouble having it spin down after the configured amount of time (it would never spin down).

If I use `hdparm` to configure the disk after connecting it, the disk would then correctly spin down after the specified amount of time. E.g. for 10 minutes:

    hdparm -S 120 -B 127 /dev/sdb

I figured out that the issue might be caused by udisks sending first the command to configure `ata-pm-standby` and second the one to configure `ata-apm-level`. For some reason the HDD only acknowledge the `ata-pm-standby` (`-S`) parameter if APM is enabled first (`-B 127`).

As an additional hint, after udisks configures the disk I can use `hdparm -S 120 /dev/sdb` (i.e. omitting the `-B` parameter from the command above) and the HDD will correctly spin down afterwards.

I could replicate this, i.e. the HDD not spinning down after having configured it, with `hdparm` by removing the udisks configuration in `/etc/udisks/` (so udisks won't try to configure the HDD when I connect it) and issuing the `hdparm` commands separately:

    hdparm -S 120 /dev/sdb; hdparm -B 127 /dev/sdb

When I issue the commands in the reverse order instead, the HDD correctly spins down after the timeout:

    hdparm -B 127 /dev/sdb; hdparm -S 120 /dev/sdb

It seems to me that the ordering of the issued commands is governed by the tests in `udiskslinuxdriveata.c`: [`if (data->ata_pm_standby != -1)`](https://github.com/storaged-project/udisks/blob/465c13438362e4e5f15ed8f802875afc6cb0c171/src/udiskslinuxdriveata.c#L1643) and [`if (data->ata_apm_level != -1)`](https://github.com/storaged-project/udisks/blob/465c13438362e4e5f15ed8f802875afc6cb0c171/src/udiskslinuxdriveata.c#L1670)

* Is the ordering of the two commands defined in some specification?
* If not, would it break other drives to invert the order of the tests to issue the commands in reverse order?
* Otherwise, would there be a workaround to make it work with the current implementation?

**Comment by crazygolem** [comment]
I updated the issue's title to make it more clear what it is about.

**Comment by tbzatek** [comment]
I vaguely remember I was having similar issues with my Toshiba MD04ACA400 drive. I don't have it anymore. What is your drive model exactly? Out of curiosity, can you please attach ouput of the `hdparm -I` command?

I couldn't find datasheet for Toshiba drives, however I suppose we can change the order the commands are sent in. Haven't found a reason not to.

Quote from the ATA8-ACS specification:

>Advanced Power Management is independent of the Standby timer setting. If both Advanced Power Management and the Standby timer are set, the device shall go to the Standby state when the timer times out or the device’s Advanced Power Management algorithm indicates that the Standby state should be entered.

No word about command ordering.

**Comment by tbzatek** [comment]
FYI, this is from the HGST He12 datasheet:
> Advanced Power Management is independent of the Standby timer setting. If both Advanced Power Management and the Standby timer are set, the device will go to the Standby state when the timer times out or the device’s Advanced Power Management algorithm indicates that the Standby state should be entered.

Still, with Ultrastar 7K6000, He6 and newer drives the APM and Standby timer settings are not honoured as expected. The drive lives its own life and I still haven't found an obvious pattern nor the way to force my settings.

**Comment by crazygolem** [comment]
So this is very strange, sometimes last week my hard drive started respecting the configured standby value (in `/etc/udisks2/<MODEL-SERIAL>.conf`), without a clear reason which would explain that.

I upgraded my system and I don't exactly remember if it was shortly before or shortly after my drive started behaving correctly. I checked the logs and `udisks` was not updated then, however `systemd` was (from 242.84 to 243.0). I'm not sure if it explains that...

I tried to reproduce the previous behavior but was unsuccessful. I don't want to downgrade my system as it is not a test system, so unfortunately I think that unless someone else is able to continue the analysis we should close the ticket :(

Anyway, here's the drive info (`hdparm -I`) for reference (without configuration from `/etc/udisks2/`):
```
ATA device, with non-removable media
	Model Number:       TOSHIBA HDWR11A                         
	Serial Number:      29N0A074FB4G
	Firmware Revision:  0603    
	Transport:          Serial, ATA8-AST, SATA 1.0a, SATA II Extensions, SATA Rev 2.5, SATA Rev 2.6, SATA Rev 3.0
Standards:
	Used: unknown (minor revision code 0x006d) 
	Supported: 10 9 8 7 6 5 
	Likely used: 10
Configuration:
	Logical		max	current
	cylinders	16383	16383
	heads		16	16
	sectors/track	63	63
	--
	CHS current addressable sectors:    16514064
	LBA    user addressable sectors:   268435455
	LBA48  user addressable sectors: 19532873728
	Logical  Sector size:                   512 bytes
	Physical Sector size:                  4096 bytes
	Logical Sector-0 offset:                  0 bytes
	device size with M = 1024*1024:     9537536 MBytes
	device size with M = 1000*1000:    10000831 MBytes (10000 GB)
	cache/buffer size  = unknown
	Form Factor: 3.5 inch
	Nominal Media Rotation Rate: 7200
Capabilities:
	LBA, IORDY(can be disabled)
	Queue depth: 32
	Standby timer values: spec'd by Standard, no device specific minimum
	R/W multiple sector transfer: Max = 16	Current = 16
	Advanced power management level: 128
	DMA: sdma0 sdma1 sdma2 mdma0 mdma1 mdma2 udma0 udma1 udma2 udma3 udma4 udma5 *udma6 
	     Cycle time: min=120ns recommended=120ns
	PIO: pio0 pio1 pio2 pio3 pio4 
	     Cycle time: no flow control=120ns  IORDY flow control=120ns
Commands/features:
	Enabled	Supported:
	   *	SMART feature set
	    	Security Mode feature set
	   *	Power Management feature set
	   *	Write cache
	   *	Look-ahead
	   *	Host Protected Area feature set
	   *	WRITE_BUFFER command
	   *	READ_BUFFER command
	   *	NOP cmd
	   *	DOWNLOAD_MICROCODE
	   *	Advanced Power Management feature set
	    	SET_MAX security extension
	   *	48-bit Address feature set
	   *	Mandatory FLUSH_CACHE
	   *	FLUSH_CACHE_EXT
	   *	SMART error logging
	   *	SMART self-test
	   *	General Purpose Logging feature set
	   *	WRITE_{DMA|MULTIPLE}_FUA_EXT
	   *	64-bit World wide name
	   *	WRITE_UNCORRECTABLE_EXT command
	   *	{READ,WRITE}_DMA_EXT_GPL commands
	   *	Segmented DOWNLOAD_MICROCODE
	    	unknown 119[6]
	    	unknown 119[7]
	   *	Gen1 signaling speed (1.5Gb/s)
	   *	Gen2 signaling speed (3.0Gb/s)
	   *	Gen3 signaling speed (6.0Gb/s)
	   *	Native Command Queueing (NCQ)
	   *	Host-initiated interface power management
	   *	Phy event counters
	   *	Host automatic Partial to Slumber transitions
	   *	Device automatic Partial to Slumber transitions
	   *	READ_LOG_DMA_EXT equivalent to READ_LOG_EXT
	    	DMA Setup Auto-Activate optimization
	    	Device-initiated interface power management
	   *	Software settings preservation
	    	unknown 78[7]
	   *	SMART Command Transport (SCT) feature set
	   *	SCT Write Same (AC2)
	   *	SCT Error Recovery Control (AC3)
	   *	SCT Features Control (AC4)
	   *	SCT Data Tables (AC5)
	   *	SANITIZE_ANTIFREEZE_LOCK_EXT command
	   *	SANITIZE feature set
	   *	OVERWRITE_EXT command
	   *	reserved 69[3]
	   *	DOWNLOAD MICROCODE DMA command
Security: 
	Master password revision code = 65534
		supported
	not	enabled
	not	locked
	not	frozen
	not	expired: security count
		supported: enhanced erase
	942min for SECURITY ERASE UNIT. 942min for ENHANCED SECURITY ERASE UNIT.
Logical Unit WWN Device Identifier: 5000039938cb126f
	NAA		: 5
	IEEE OUI	: 000039
	Unique ID	: 938cb126f
Checksum: correct
```

**Comment by tbzatek** [comment]
FYI, the HGST Ultrastar He12 SATA datasheet: https://documents.westerndigital.com/content/dam/doc-library/en_us/assets/public/western-digital/product/data-center-drives/ultrastar-dc-hc500-series/product-manual-ultrastar-dc-hc520-sata-oem-spec.pdf

Another interesting reading about _Background Medium Scan_, only partially related: https://forum.qnap.com/viewtopic.php?t=131537

**Comment by crazygolem** [comment]
Interesting, that BMS feature. I'll try to have a look next time I have a drive that presents a similar issue. I don't think however that it is related:
* The post mentions that Hitachi/WD said that BMS should not prevent spindown (might not apply to any other vendor/device though)
* If it were related, reordering the commands with hdparm should not have solved the problem imo.

**Comment by tbzatek** [comment]
The _Background Medium Scan_ feature appears to be SAS-only specific, though SATA drives may implement similar feature in their firmware, it doesn't seem to be host-controllable. And I suppose it's highly vendor-specific anyway.

I've posted a patch reordering setting of the two flags, it wouldn't harm as I haven't found indication of a specific command order requirement.

FYI, I've also discovered the _Extended Power Conditions (EPC)_ feature that controls drive power management in a similar way to _Advanced Power Management (APM)_, only with finer granularity and persistency. It appears to be mutually exclusive with each other however and I guess there's no point of having both implemented in `udisks`, at the current stage.
